### PR TITLE
Use Kokkos in functions struct replacing TBB

### PR DESCRIPTION
**Pull request by masterleinad**
Currently, we are using TBB for most vector operations on the host and `Kokkos` for the `Default` memory space. This pull request unifies the implementation by always using `Kokkos`. We should evaluate if the performance on the host is acceptable (if `Kokkos` was configured with a parallel host execution space).

### CODE REVIEW

**Comment by tamiko** [comment]
@masterleinad @kronbichler What would be a good setup to test this? I am happy to run some extended performance tests.

**Comment by tjhei** [comment]
> if `Kokkos` was configured with a parallel host execution space).

If I configure with Kokkos_ENABLE_THREADS or Kokkos_ENABLE_OPENMP, right? Any preference between them or anything else we need to know?

It would also be sensible to check if we see no overhead for an MPI-only run where one would not want vector operations to be threaded...

**Comment by Rombur** [comment]
> If I configure with Kokkos_ENABLE_THREADS or Kokkos_ENABLE_OPENMP, right? Any preference between them or anything else we need to know?

Probably better to use OpenMP over threads. You want to set `OMP_PROC_BIND=spread` and `OMP_PLACES=threads` to get better performance.

**Comment by masterleinad** [comment]
> @masterleinad @kronbichler What would be a good setup to test this? I am happy to run some extended performance tests.

The benchmark in https://github.com/dealii/dealii/issues/2496#issuecomment-208233298 using `LinearAlgebra::distributed::Vector` would likely be suitable. Up to now, `LinearAlgebra::Vector` is not affected by this pull request.

File: include/deal.II/base/memory_space_data.h [hunk 1]
```diff
@@ -113,7 +116,12 @@ namespace MemorySpace
         0))
   {}
 
-
+  template <typename T, typename MemorySpace>
+  MemorySpaceData<T, MemorySpace>::MemorySpaceData(
+    Kokkos::View<T *, typename MemorySpace::kokkos_space> new_values)
+    : values_host_buffer(Kokkos::View<T *, Kokkos::HostSpace>("host buffer", 0))
+    , values(new_values)
```
Comment (bangerth) [inline include/deal.II/base/memory_space_data.h:123]: You're just copying `new_values` twice this way. Make it a `const` reference to avoid one of them.

Comment (masterleinad) [reply include/deal.II/base/memory_space_data.h:123]: I don't need this constructor anymore anyway. It probably makes sense to replace `MemorySpaceData` with `Kokkos::View` at least internally in most places.

**Comment by bangerth** [comment]
@masterleinad What do you want to do with this PR?

**Comment by masterleinad** [comment]
> @masterleinad What do you want to do with this PR?

I don't have plans to pursue this further at the moment. The performance results weren't that convincing when I initially opened this pull request.

**Pull request closed without merging**